From c3126c59ec0a78e0f8ff311fcfe1ee87516167b9 Mon Sep 17 00:00:00 2001
From: Sven Neumann <neumann@teufel.de>
Date: Wed, 30 Sep 2015 21:21:12 +0200
Subject: [PATCH 2/3] connman: cancel a pending g_dbus_proxy_new_for_bus() call

If a service is removed while the g_dbus_proxy_new_for_bus()
call is pending, then this call needs to be cancelled.
---
 libgupnp/gupnp-connman-manager.c | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/libgupnp/gupnp-connman-manager.c b/libgupnp/gupnp-connman-manager.c
index a764c80..50da0ed 100644
--- a/libgupnp/gupnp-connman-manager.c
+++ b/libgupnp/gupnp-connman-manager.c
@@ -44,6 +44,7 @@ typedef struct {
         GUPnPConnmanManager *manager;
         GUPnPContext        *context;
         GDBusProxy          *proxy;
+        GCancellable        *cancellable;
         CMServiceState      current;
         guint               sig_prop_id;
         guint               port;
@@ -248,6 +249,14 @@ cm_service_free (CMService *cm_service)
 
                 g_object_unref (cm_service->proxy);
         }
+        else if (cm_service->cancellable != NULL)  {
+                g_cancellable_cancel (cm_service->cancellable);
+        }
+
+        if (cm_service->cancellable != NULL)  {
+                g_object_unref (cm_service->cancellable);
+                cm_service->cancellable = NULL;
+        }
 
         if (cm_service->context != NULL) {
                 g_signal_emit_by_name (cm_service->manager,
@@ -357,7 +366,8 @@ service_proxy_new_cb (GObject      *source_object,
         service_proxy = g_dbus_proxy_new_for_bus_finish (res, &error);
 
         if (error != NULL) {
-                g_warning ("Failed to create D-Bus proxy: %s", error->message);
+                if (!g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+                        g_warning ("Failed to create D-Bus proxy: %s", error->message);
                 g_error_free (error);
 
                 return;
@@ -420,6 +430,8 @@ cm_service_add (GUPnPConnmanManager *manager,
                              g_strdup (path),
                              cm_service);
 
+        cm_service->cancellable = g_cancellable_new ();
+
         g_dbus_proxy_new_for_bus (G_BUS_TYPE_SYSTEM,
                                   G_DBUS_PROXY_FLAGS_DO_NOT_LOAD_PROPERTIES |
                                   G_DBUS_PROXY_FLAGS_DO_NOT_CONNECT_SIGNALS,
@@ -427,7 +439,7 @@ cm_service_add (GUPnPConnmanManager *manager,
                                   CM_DBUS_CONNMAN_NAME,
                                   path,
                                   CM_DBUS_SERVICE_INTERFACE,
-                                  NULL,
+                                  cm_service->cancellable,
                                   service_proxy_new_cb,
                                   manager);
 }
-- 
2.4.3

